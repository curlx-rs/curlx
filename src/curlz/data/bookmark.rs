use serde::{Deserialize, Serialize};

use crate::data::HttpRequest;
use crate::ops::SaveBookmark;

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct Bookmark {
    pub slug: String,
    pub request: HttpRequest,
}

impl AsRef<HttpRequest> for Bookmark {
    fn as_ref(&self) -> &HttpRequest {
        &self.request
    }
}

impl AsRef<str> for Bookmark {
    fn as_ref(&self) -> &str {
        self.slug.as_ref()
    }
}

impl Bookmark {
    pub fn slug(&self) -> &str {
        self.as_ref()
    }

    pub fn request(&self) -> &HttpRequest {
        self.as_ref()
    }
}

impl<'a> From<&SaveBookmark<'a>> for Bookmark {
    fn from(cmd: &SaveBookmark) -> Self {
        Self {
            slug: cmd.slug.clone(),
            request: cmd.bookmark.clone(),
        }
    }
}
