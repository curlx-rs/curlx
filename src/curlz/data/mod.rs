mod bookmark;
mod request;

pub use bookmark::*;
pub use request::*;
