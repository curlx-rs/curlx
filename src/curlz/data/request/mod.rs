mod http_headers;
mod http_method;
mod http_request;
mod http_version;

pub use http_headers::*;
pub use http_method::*;
pub use http_request::*;
pub use http_version::*;
