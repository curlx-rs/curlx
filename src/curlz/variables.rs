use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
pub struct Placeholder {
    pub name: String,
    pub value: Option<String>,
    pub default: Option<String>,
    pub prompt: Option<String>,
}

impl Placeholder {
    pub fn new(key: impl AsRef<str>, value: impl AsRef<str>) -> Self {
        Self {
            name: key.as_ref().to_owned(),
            value: value.as_ref().to_owned().into(),
            default: None,
            prompt: None,
        }
    }
}
