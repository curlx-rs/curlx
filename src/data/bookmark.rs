use serde::{Deserialize, Serialize};

use crate::{HttpRequest, SaveBookmarkCommand};

#[derive(Debug, Serialize, Deserialize)]
pub struct Bookmark {
    slug: String,
    request: HttpRequest,
}

impl AsRef<HttpRequest> for Bookmark {
    fn as_ref(&self) -> &HttpRequest {
        &self.request
    }
}

impl AsRef<str> for Bookmark {
    fn as_ref(&self) -> &str {
        self.slug.as_ref()
    }
}

impl Bookmark {
    pub fn slug(&self) -> &str {
        self.as_ref()
    }

    pub fn request(&self) -> &HttpRequest {
        self.as_ref()
    }
}

impl<'a> From<&SaveBookmarkCommand<'a>> for Bookmark {
    fn from(cmd: &SaveBookmarkCommand) -> Self {
        Self {
            slug: cmd.slug.clone(),
            request: cmd.to_bookmark.clone(),
        }
    }
}
