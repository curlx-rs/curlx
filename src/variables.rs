use regex::Regex;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct TemplateSlots {
    pub(crate) var_name: String,
    pub(crate) var_info: VarInfo,
    pub(crate) prompt: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum VarInfo {
    Bool { default: Option<bool> },
    String { entry: Box<StringEntry> },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StringEntry {
    pub(crate) default: Option<String>,
    pub(crate) choices: Option<Vec<String>>,
    #[serde(with = "serde_regex")]
    pub(crate) regex: Option<Regex>,
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
enum VarDataType {
    Bool,
    String,
}
