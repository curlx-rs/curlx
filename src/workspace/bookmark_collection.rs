use anyhow::anyhow;
use convert_case::{Case, Casing};
use filenamify::filenamify;
use std::path::PathBuf;
use std::{env, fs};

use crate::data::Bookmark;
use crate::Result;

const WORKSPACE_FOLDER: &str = ".curlx";
const BOOKMARK_FOLDER: &str = "bookmarks";

pub struct BookmarkCollection {
    working_dir: PathBuf,
}

impl BookmarkCollection {
    pub fn new() -> Result<Self> {
        Ok(Self {
            working_dir: env::current_dir()
                .map_err(|e| anyhow!("cannot create processor: {}", e))?,
        })
    }
}

impl BookmarkCollection {
    pub fn save(&self, bookmark: &Bookmark) -> Result<()> {
        let slug = bookmark.slug();
        let request = bookmark.request();

        let file_name = filenamify(format!("{:?} {}", &request.method, slug)).to_case(Case::Snake);
        let bookmark = serde_yaml::to_string(&bookmark)?;

        let bookmarks_path = self
            .working_dir
            .join(WORKSPACE_FOLDER)
            .join(BOOKMARK_FOLDER);
        fs::create_dir_all(bookmarks_path.as_path())?;
        {
            fs::write(
                bookmarks_path.join(format!("{}.yml", file_name.as_str())),
                bookmark,
            )
            .map_err(|e| anyhow!("cannot write request bookmark to file: {}", e))
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::data::HttpMethod;
    use crate::variables::{StringEntry, VarInfo};
    use crate::{HttpHeaders, HttpRequest, SaveBookmarkCommand, TemplateSlots};
    use regex::Regex;
    use tempfile::{tempdir, TempDir};

    impl BookmarkCollection {
        pub fn temporary() -> (Self, TempDir) {
            let tempdir = tempdir().unwrap();
            (
                Self {
                    working_dir: tempdir.path().to_path_buf(),
                },
                tempdir,
            )
        }
    }

    #[test]
    fn should_handle_save_bookmark_command() {
        let request = HttpRequest {
            url: "{{protonmail_api_baseurl}}/pks/lookup?op=get&search={{email}}".to_owned(),
            method: HttpMethod::Get,
            headers: HttpHeaders::default(),
            curl_params: vec![],
            placeholders: vec![email_placeholder(), protonmail_api_baseurl_placeholder()],
        };
        let cmd = SaveBookmarkCommand::new("/protonmail/gpg/:email", &request);

        let (p, tmp) = BookmarkCollection::temporary();
        p.save(&(&cmd).into()).unwrap();

        let saved_bookmark = String::from_utf8(
            fs::read(
                tmp.path()
                    .join(WORKSPACE_FOLDER)
                    .join(BOOKMARK_FOLDER)
                    .join("get_protonmail_gpg_email.yml"),
            )
            .unwrap(),
        )
        .unwrap();

        insta::assert_snapshot!(saved_bookmark);
    }

    fn email_placeholder() -> TemplateSlots {
        TemplateSlots {
            var_name: "email".to_string(),
            var_info: VarInfo::String {
                entry: Box::new(StringEntry {
                    default: None,
                    choices: None,
                    regex: Some(
                        Regex::new(r"^([a-zA-Z0-9_\-\.]+)@([a-zA-Z0-9_\-\.]+)\.([a-zA-Z]{2,5})$")
                            .unwrap(),
                    ),
                }),
            },
            prompt: "enter an email address".to_string(),
        }
    }

    fn protonmail_api_baseurl_placeholder() -> TemplateSlots {
        TemplateSlots {
            var_name: "protonmail_api_baseurl".to_string(),
            var_info: VarInfo::String {
                entry: Box::new(StringEntry {
                    default: Some("https://api.protonmail.ch".to_string()),
                    choices: None,
                    regex: None,
                }),
            },
            prompt: "enter the protonmail api baseurl".to_string(),
        }
    }
}
