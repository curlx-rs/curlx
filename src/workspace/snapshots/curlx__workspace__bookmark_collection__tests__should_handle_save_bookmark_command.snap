---
source: src/workspace/bookmark_collection.rs
expression: saved_bookmark
---
---
slug: "/protonmail/gpg/:email"
request:
  url: "{{protonmail_api_baseurl}}/pks/lookup?op=get&search={{email}}"
  method: GET
  headers: []
  curl_params: []
  placeholders:
    - var_name: email
      var_info:
        String:
          entry:
            default: ~
            choices: ~
            regex: "^([a-zA-Z0-9_\\-\\.]+)@([a-zA-Z0-9_\\-\\.]+)\\.([a-zA-Z]{2,5})$"
      prompt: enter an email address
    - var_name: protonmail_api_baseurl
      var_info:
        String:
          entry:
            default: "https://api.protonmail.ch"
            choices: ~
            regex: ~
      prompt: enter the protonmail api baseurl

